docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,83 @@
+# Storage Automatic Scaler
+
+## Summary
+
+This feature would help users to get rid of maintenance task of keeping the system with enough available free space.
+In virtualized environments, we can scale the storage devices Osds so this feature helps in automatic scaling. 
+
+## Proposal
+
+Implement auto storage scale CR per device class of a specific Storagecluster.
+
+## High-Level Design
+
+- Create a StorageAutoScale CR per Storagecluster device class
+    ```
+    apiVersion: ocs.openshift.io/v1
+    kind: StorageAutoScaling
+    metadata:
+        name: <storageclusterName>-<deviceClassName>
+        namespace: openshift-storage
+    spec:
+        storageCluster: 
+            name: <storageclusterName>
+        deviceClassName: ssd
+        clusterCapacityLimit: <capacity>
+        maxOsdSize: 8Ti
+        timeoutSeconds: 1800   
+    status:
+        phase: (NotStarted | InProgress | Failed | Succeeded)
+        LastRunTimestamp: (only on Failed | Succeeded)
+        currentOsdSize: 2TiB
+        expectedOsdSize: 4TiB
+        currentOsdCount: 3​
+        expectedOsdCount: 3
+    ```
+
+- Watch prometheus periodically every 10 mins and depending on the Osd fill percentage, reconcile the StorageAutoScaling CR.
+  
+- If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+
+- If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of 8Tib on each storage node, by simply increasing the (count+1) in the device set associated with a device class.   
+
+## Low-Level Design 
+
+1) Create a controller that requeueAfter(https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Result) every 10 mins.
+   
+2) List all the device class Osds used percentage using Prometheus metrics.
+    Query: (ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))
+       `OsdPercentage`:
+        ```
+        osd1-ssd               69%
+        osd2-ssd               72%
+        ```
+
+3) If the highest osd percentage reported is more than 70%(reaching osd nearfull), scaling is needed.
+
+4) If scaling is needed calculate the `expectedOsdSize` and `expectedOsdCount`.
+    1)  Calculate the total number osds and osd size from the prometheus response.
+    2)  Calculate the total number osds and osd size from the storageCluster.
+    3)  If the above values are equal(needed to avoid double re-sizing), create or update the `OsdPercentage map`
+        1)  If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+        2)  If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of maxOsdSize(default:8Tib) on each storage node.
```

Comment:
The OSDs will be added for each device set, right? It might be per-zone, rack, or host.
```suggestion
        2)  If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of maxOsdSize(default:8Tib) on each `storageDeviceSet`
```

Reply:
correct